src/libraries/System.Net.Sockets/src/System/Net/Connections/SocketsConnectionFactory.cs

```diff
@@ -0,0 +1,159 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.IO;
+using System.IO.Pipelines;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Connections
+{
+    /// <summary>
+    /// A <see cref="ConnectionFactory"/> to establish socket-based connections.
+    /// </summary>
+    /// <remarks>
+    /// When constructed with <see cref="ProtocolType.Tcp"/>, this factory will create connections with <see cref="Socket.NoDelay"/> enabled.
+    /// In case of IPv6 sockets <see cref="Socket.DualMode"/> is also enabled.
+    /// </remarks>
+    public class SocketsConnectionFactory : ConnectionFactory, SocketConnection.IDataChannelProvider
+    {
+        private readonly AddressFamily _addressFamily;
+        private readonly SocketType _socketType;
+        private readonly ProtocolType _protocolType;
+
+        // use same message as the default ctor
+        private static readonly string s_cancellationMessage = new OperationCanceledException().Message;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SocketsConnectionFactory"/> class.
+        /// </summary>
+        /// <param name="addressFamily">The <see cref="AddressFamily"/> to forward to the socket.</param>
+        /// <param name="socketType">The <see cref="SocketType"/> to forward to the socket.</param>
+        /// <param name="protocolType">The <see cref="ProtocolType"/> to forward to the socket.</param>
+        public SocketsConnectionFactory(
+            AddressFamily addressFamily,
+            SocketType socketType,
+            ProtocolType protocolType)
+        {
+            _addressFamily = addressFamily;
+            _socketType = socketType;
+            _protocolType = protocolType;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SocketsConnectionFactory"/> class
+        /// that will forward <see cref="AddressFamily.InterNetworkV6"/> to the Socket constructor.
+        /// </summary>
+        /// <param name="socketType">The <see cref="SocketType"/> to forward to the socket.</param>
+        /// <param name="protocolType">The <see cref="ProtocolType"/> to forward to the socket.</param>
+        /// <remarks>The created socket will be an IPv6 socket with <see cref="Socket.DualMode"/> enabled.</remarks>
+        public SocketsConnectionFactory(SocketType socketType, ProtocolType protocolType)
+            : this(AddressFamily.InterNetworkV6, socketType, protocolType)
+        {
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<Connection> ConnectAsync(
```

Comment:
The previous type has `EstablishConnectionAsync` with a HttpRequestMessage argument. Is the following snippet how you would replace that...

```cs
public override ValueTask<Conneciton> ConnectAsync(
    EndPoint? endPoint,
    IConnectionProperties? options = null,
    CancellationToken cancellationToken = default)
{
    if (options.TryGet(typeof(HttpRequesMessage), out var message))
    {
        // ...
    }

    return base.ConnectAsync(endpoint, options, cancellationToken);
}
```
I didn't see a test that attempts that. Consider adding a test for this scenario to ensure it doesn't accidentally regress.